### PR TITLE
Tracing: populate tags based on the custom headers

### PR DESCRIPTION
**Pull request by RomanDzhabarov**

### CODE REVIEW

File: docs/configuration/http_conn_man/tracing.rst [hunk 1]
```diff
@@ -7,12 +7,18 @@ Tracing
...
+      "request_headers_for_tags": []
     }
   }
  
 operation_name
   *(required, string)* Span name will be derived from operation_name. "ingress" and "egress"
   are the only supported values.
 
+request_headers_for_tags
+  *(optional, array)* An optional list of header names which is used for populating tags the an active span.
```
Comment (ccaraman) [inline docs/configuration/http_conn_man/tracing.rst]: last part of the sentence doesn't make sense.

Comment (RomanDzhabarov) [inline docs/configuration/http_conn_man/tracing.rst]: yeah, 'on' is missing

File: docs/configuration/http_conn_man/tracing.rst [hunk 2]
```diff
@@ -7,12 +7,18 @@ Tracing
...
+      "request_headers_for_tags": []
     }
   }
  
 operation_name
   *(required, string)* Span name will be derived from operation_name. "ingress" and "egress"
   are the only supported values.
 
+request_headers_for_tags
+  *(optional, array)* An optional list of header names which is used for populating tags on the an active span.
```
Comment (mattklein123) [inline docs/configuration/http_conn_man/tracing.rst]: typo

File: docs/configuration/http_conn_man/tracing.rst [hunk 3]
```diff
@@ -7,12 +7,18 @@ Tracing
...
+  Each tag name is the header name and tag value is the header value from the request headers.
+  If specified header is not present in the request headers no tag is created.
```
Comment (mattklein123) [inline docs/configuration/http_conn_man/tracing.rst]: grammar

File: include/envoy/http/header_map.h [hunk 4]
```diff
@@ -22,6 +22,10 @@ class LowerCaseString {
 
   const std::string& get() const { return string_; }
   bool operator==(const LowerCaseString& rhs) const { return string_ == rhs.string_; }
+  LowerCaseString& operator=(const LowerCaseString& rhs) {
```
Comment (mattklein123) [inline include/envoy/http/header_map.h]: should not be needed and is probably indicative of doing something wrong.

File: include/envoy/tracing/http_tracer.h [hunk 5]
```diff
@@ -15,7 +15,15 @@ class Config {
 public:
   virtual ~Config() {}
 
+  /**
+   * Operation name for tracing, e.g., ingress.
```
Comment (mattklein123) [inline include/envoy/tracing/http_tracer.h]: @return

File: include/envoy/tracing/http_tracer.h [hunk 6]
```diff
@@ -15,7 +15,15 @@ class Config {
...
+   */
   virtual OperationName operationName() const PURE;
+
+  /**
+   * List of headers to populate tags on the active span.
```
Comment (mattklein123) [inline include/envoy/tracing/http_tracer.h]: @return

File: source/common/http/conn_manager_impl.cc [hunk 7]
```diff
@@ -315,6 +315,8 @@ ConnectionManagerImpl::ActiveStream::~ActiveStream() {
     if (request_info_.healthCheck()) {
       connection_manager_.config_.tracingStats().health_check_.inc();
     } else {
+      Tracing::HttpTracerUtility::populateTagsBasedOnHeaders(*active_span_, *request_headers_,
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.cc]: just move this logic into finalizeSpan

File: source/common/http/conn_manager_impl.h [hunk 8]
```diff
@@ -104,6 +104,7 @@ struct ConnectionManagerTracingStats {
  */
 struct TracingConnectionManagerConfig {
   Tracing::OperationName operation_name_;
+  std::list<Http::LowerCaseString> request_headers_for_tags_;
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.h]: std::vector

File: include/envoy/http/header_map.h (hunk 4 above)
Comment (RomanDzhabarov) [inline include/envoy/http/header_map.h]: that's because Optional does not have a support for moveable types. Would you suggest to add it there?

Comment (mattklein123) [inline include/envoy/http/header_map.h]: I don't know what the actual problem you are trying to solve is, but I think you shouldn't need any of it, so unclear on the actual problem.

Comment (RomanDzhabarov) [inline include/envoy/http/header_map.h]: as far as i can tell the following happens:
Optional.value(arg) method does copy assignment for the argument passed in, what ends up happening vector<LowerCaseString> is copy assigned and consequently LowerCaseString is copy assigned which is not permitted as we have move ctor for the LowerCaseString and compiler removes implicit operator=(const LowerCaseString&).

That's why one solution is to explicitly define copy assignment.

Comment (mattklein123) [inline include/envoy/http/header_map.h]: per offline convo make config a pointer to a struct and remove operator= from here (so we don't introduce accidental copies).

File: docs/configuration/http_conn_man/tracing.rst [hunk 9]
```diff
@@ -7,12 +7,18 @@ Tracing
...
+      "request_headers_for_tags": []
     }
   }
  
 operation_name
   *(required, string)* Span name will be derived from operation_name. "ingress" and "egress"
   are the only supported values.
 
+request_headers_for_tags
+  *(optional, array)* An optional list of header names which is used for populating tags on the active span.
```
Comment (mattklein123) [inline docs/configuration/http_conn_man/tracing.rst]: still multiple grammar issues. Please pair write with someone.

Comment (RomanDzhabarov) [inline docs/configuration/http_conn_man/tracing.rst]: will do tomorrow morning

**Pull request merged**